keps/sig-autoscaling/0000-kep-template.md

```diff
@@ -0,0 +1,273 @@
+will get allocated and thus how many new nodes are needed to scale up.
+
+Each of the mentioned cloud providers has its own implementation of constructing
+the node template with its own list of information pulled from cloud provider API:
+
+* Azure:
+  - instance type to determine resource capacity (CPU, GPU, Memory)
+  - instance tags to construct node labels
+  - generic node labels: Arch, OS, instance type, region, failure domain, hostname
+* AWS:
+  - instance type to determine resource capacity (CPU, GPU, Memory)
+  - instance tags to construct node labels and taints
+  - generic node labels: Arch, OS, instance type, region, failure domain, hostname
+* GCE:
+  - instance type to determine resource capacity (CPU, Memory)
+  - instance tags to construct node labels (e.g. gpu.ResourceNvidiaGPU)
+  - maximum number of pods: apiv1.ResourcePods
+  - taints (from instance metadata)
+  - kubelet system and reserved allocatable
+
+## Code status
+
+At this point the cluster autoscaler does not support node autoprovisioning (TODO(jchaloup): link to the proposal).
+Every implementation assumes every node group already exists in the cloud provider.
+Both `Create` and `Delete` operations return an error `ErrAlreadyExist` and the autoprovisioning is off.
+
+The autoscaler allows to either enumerate all node groups (via `--nodes` option)
+or automatically discover all node groups on-the-fly (via `--node-group-auto-discovery`).
+
+## Integration with machine API
+
+Given the scaling up/down operation corresponds to increasing/decreasing the number
+of nodes in a node group, it's sufficient to "just" change the number of replicas
+of corresponding machine set.
+
+### Node template
+
+The `machine` object holds all cloud provider specific information under
+`spec.providerConfig` field. Given only the actuator can interpret the provider
+config and there can be multiple actuator implementations for the same cloud provider,
+all information necessary to render the template must be contained in the machine spec (outside
+the provider config). This may lead to data duplication as one needs to provide
+a machine type in the config (e.g. to provision an instance) and cpu, gpu and memory
+requirements of the machine outside the config (e.g. to construct node object
+resource capacity).
+
+Implemented through labels the `machine` spec can look like:
+
+```yaml
+---
+apiVersion: cluster.k8s.io/v1alpha1
+kind: Machine
+metadata:
+  name: node-group-machine
+  namespace: application
+  labels:
+    sigs.k8s.io/cluster-autoscaler-resource-capacity: "cpu:200m,gpu:1,memory:4GB"
+    sigs.k8s.io/cluster-autoscaler-region: us-west-1
+...
+```
+
+Other option is to pull all the information from the `spec.providerConfig`.
+Which means the autoscaler needs to import actuator specific code, which forces
+to choose a single implementation of actuator for each cloud provider.
+That can be to restrictive.
```

Comment:
Dependency management in autoscaler is already very hard. I don't really like the idea of adding a whole bunch of implementations of cluster API.

Reply:
Agreed. I would also expect that machine-api should have necessary generic details ready at MachineClass or MachineSet/Deployment, which can be readily consumed by autoscaler without knowing much about the actuator-specific apis.

Reply:
Agreed. I would also expect that machine-api should have necessary generic details ready at MachineClass or MachineSet/Deployment, which can be readily consumed by autoscaler without knowing much about the actuator-specific apis.